src/plugins/explore/public/components/visualizations/table/table_vis.tsx

```diff
@@ -100,6 +105,33 @@ export const TableVis = React.memo(
       }
     }, [styleOptions?.showColumnFilter]);
 
+    // Handle column visibility when columns change
+    // If there are no common fields between old and new columns, show all new columns
+    // If there are common fields, preserve existing visible columns and add any new ones
+    useEffect(() => {
+      const currentFieldNames = sortedColumns.map(({ name }) => name);
+      const prevFieldNames = prevFieldNamesRef.current;
```

Comment:
I prefer to delete this ref, can we calculate the new columns based `prevVisibleColumns` in `setVisibleColumns` callback and `sortedColumns`?